File: solutions/haskell/bob/3/src/Bob.hs
module Bob (responseFor) where

import Data.Char
import Data.List

responseFor :: String -> String
responseFor input
  | silence = "Fine. Be that way!"
  | question && yelling = "Calm down, I know what I'm doing!"
  | question = "Sure."
  | yelling = "Whoa, chill out!"
  | otherwise = "Whatever."
  where
    chars = filter (not . isSpace) input
    question = "?" `isSuffixOf` chars
    yelling = (not . any isLower) chars && any isAlpha chars
    silence = null chars

File: solutions/elixir/bob/3/lib/bob.ex
defmodule Bob do
  defp yelling?(input) do
    String.upcase(input) == input && String.downcase(input) !== input
  end

  defp question?(input) do
    String.ends_with?(input, "?")
  end

  defp silence?(input) do
    String.replace(input, ~r/[\r\t\n]/, "") == ""
  end

  def hey(input) do
    input = String.trim(input)

    cond do
      silence?(input) -> "Fine. Be that way!"
      yelling?(input) && question?(input) -> "Calm down, I know what I'm doing!"
      yelling?(input) -> "Whoa, chill out!"
      question?(input) -> "Sure."
      true -> "Whatever."
    end
  end
end
